File: FBSimulatorControl/Configuration/FBSimulatorBootConfiguration.m
/*
 * Copyright (c) Meta Platforms, Inc. and affiliates.
 *
 * This source code is licensed under the MIT license found in the
 * LICENSE file in the root directory of this source tree.
 */

#import "FBSimulatorBootConfiguration.h"

#import "FBSimulator.h"
#import "FBSimulatorError.h"

@implementation FBSimulatorBootConfiguration

#pragma mark Initializers

+ (instancetype)defaultConfiguration
{
  static dispatch_once_t onceToken;
  static FBSimulatorBootConfiguration *configuration;
  dispatch_once(&onceToken, ^{
    configuration = [[FBSimulatorBootConfiguration alloc] initWithOptions:FBSimulatorBootOptionsVerifyUsable environment:@{}];
  });
  return configuration;
}

- (instancetype)initWithOptions:(FBSimulatorBootOptions)options environment:(NSDictionary<NSString *, NSString *> *)environment
{
  self = [super init];
  if (!self) {
    return nil;
  }

  _options = options;
  _environment = environment;

  return self;
}

#pragma mark NSCopying

- (instancetype)copyWithZone:(NSZone *)zone
{
  // Instances are immutable.
  return self;
}

#pragma mark NSObject

- (BOOL)isEqual:(FBSimulatorBootConfiguration *)configuration
{
  if (![configuration isKindOfClass:self.class]) {
    return NO;
  }

  return self.options == configuration.options &&
    (self.environment == configuration.environment || [self.environment isEqualToDictionary:configuration.environment]);
}
  
- (NSUInteger)hash
{
  return self.options ^ self.environment.hash;
}

- (NSString *)description
{
  return [NSString stringWithFormat:
    @"Boot Environment %@ | Options %@",
    [FBCollectionInformation oneLineDescriptionFromDictionary:self.environment],
    [FBCollectionInformation oneLineDescriptionFromArray:[FBSimulatorBootConfiguration stringsFromBootOptions:self.options]]
  ];
}

#pragma mark Utility

static NSString *const BootOptionStringDirectLaunch = @"Direct Launch";

+ (NSArray<NSString *> *)stringsFromBootOptions:(FBSimulatorBootOptions)options
{
  NSMutableArray<NSString *> *strings = [NSMutableArray array];
  if ((options & FBSimulatorBootOptionsTieToProcessLifecycle) == FBSimulatorBootOptionsTieToProcessLifecycle) {
    [strings addObject:BootOptionStringDirectLaunch];
  }
  return [strings copy];
}

@end

File: FBSimulatorControl/Configuration/FBSimulatorBootConfiguration.swift
/*
 * Copyright (c) Meta Platforms, Inc. and affiliates.
 *
 * This source code is licensed under the MIT license found in the
 * LICENSE file in the root directory of this source tree.
 */

import Foundation

/// An Option Set for Direct Launching.
///
/// The raw values are not contiguous, and are kept as they were when this was an `NS_OPTIONS`:
/// they are compared and persisted by callers.
public struct FBSimulatorBootOptions: OptionSet, Sendable {

  public let rawValue: UInt

  public init(rawValue: UInt) {
    self.rawValue = rawValue
  }

  /// Ties the Simulator's lifecycle to that of the launching process, so that the Simulator is
  /// shut down automatically when the process that booted it dies.
  public static let tieToProcessLifecycle = FBSimulatorBootOptions(rawValue: 1 << 1)

  /// Requires that the Simulator is 'Usable' before the boot API completes. A Simulator can report
  /// itself 'Booted' very quickly while not yet being usable.
  public static let verifyUsable = FBSimulatorBootOptions(rawValue: 1 << 3)
}

@objc(FBSimulatorBootConfiguration)
public class FBSimulatorBootConfiguration: NSObject, NSCopying {

  // MARK: Properties

  public let options: FBSimulatorBootOptions

  @objc public let environment: [String: String]

  // MARK: Default Instance

  @objc(defaultConfiguration)
  public nonisolated(unsafe) static let `default` = FBSimulatorBootConfiguration(
    options: .verifyUsable,
    environment: [:]
  )

  // MARK: Initializers

  public init(options: FBSimulatorBootOptions, environment: [String: String]) {
    self.options = options
    self.environment = environment
    super.init()
  }

  // MARK: - NSCopying

  public func copy(with zone: NSZone? = nil) -> Any {
    self
  }

  // MARK: - NSObject

  public override func isEqual(_ object: Any?) -> Bool {
    guard let other = object as? FBSimulatorBootConfiguration else {
      return false
    }
    return options == other.options && environment == other.environment
  }

  public override var hash: Int {
    Int(bitPattern: options.rawValue) ^ (environment as NSDictionary).hash
  }

  public override var description: String {
    String(
      format: "Boot Environment %@ | Options %@",
      FBCollectionInformation.oneLineDescription(from: environment as [String: Any]),
      FBCollectionInformation.oneLineDescription(from: Self.stringsFromBootOptions(options) as [Any])
    )
  }

  // MARK: - Private

  private static let bootOptionStringDirectLaunch = "Direct Launch"

  private static func stringsFromBootOptions(_ options: FBSimulatorBootOptions) -> [String] {
    var strings: [String] = []
    if options.contains(.tieToProcessLifecycle) {
      strings.append(bootOptionStringDirectLaunch)
    }
    return strings
  }
}
